cmd/api/src/api/v2/ad_related_entity.go

```diff
@@ -0,0 +1,405 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package v2
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	adAnalysis "github.com/specterops/bloodhound/analysis/ad"
+	"github.com/specterops/bloodhound/dawgs/graph"
+	"github.com/specterops/bloodhound/dawgs/ops"
+	"github.com/specterops/bloodhound/graphschema/ad"
+	"github.com/specterops/bloodhound/src/api"
+	"github.com/specterops/bloodhound/src/model/appcfg"
+	"github.com/specterops/bloodhound/src/queries"
+)
+
+// handleAdRelatedEntityQuery is for retrieving entities related to a specific entity.
+// This endpoint returns a polymorphic response dependent upon which return type is requested
+// by the `type` parameter, which can be `list`, `count`, or `graph`.
+// Path delegates are for graphing, list delegates are for listing and counting. Endpoints
+// without a certain delegate do not support that delegate feature.
+func (s *Resources) handleAdRelatedEntityQuery(response http.ResponseWriter, request *http.Request, queryName string, pathDelegate any, listDelegate any) {
+	if params, err := queries.BuildEntityQueryParams(request, queryName, pathDelegate, listDelegate); err != nil {
+		api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, fmt.Sprintf(api.FmtErrorResponseDetailsBadQueryParameters, err), request), response)
+	} else if entityPanelCachingFlag, err := s.DB.GetFlagByKey(appcfg.FeatureEntityPanelCaching); err != nil {
+		api.HandleDatabaseError(request, response, err)
+	} else if results, err := s.GraphQuery.GetADEntityQueryResult(request.Context(), params, entityPanelCachingFlag.Enabled); err != nil {
+		if errors.Is(err, queries.ErrGraphUnsupported) || errors.Is(err, queries.ErrUnsupportedDataType) {
+			api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, fmt.Sprintf(api.FmtErrorResponseDetailsBadQueryParameters, err), request), response)
+		} else if errors.Is(err, ops.ErrTraversalMemoryLimit) {
+			api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusInternalServerError, "calculating the request results exceeded memory limitations due to the volume of objects involved", request), response)
+		} else {
+			api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusInternalServerError, "an unknown error occurred during the request", request), response)
+		}
+	} else {
+		api.WriteJSONResponse(request.Context(), results, http.StatusOK, response)
+	}
+}
+
+func (s *Resources) ListADUserSessions(response http.ResponseWriter, request *http.Request) {
+	var (
+		queryName    = "ListADUserSessions"
+		pathDelegate = adAnalysis.FetchUserSessionPaths
+		listDelegate = adAnalysis.FetchUserSessions
```

Comment:
@zinic this also seems like a good place to cut out 3 allocs, so i'll make the same change on these